sdk/src/main/scala/io/horizen/account/sc2sc/AbstractCrossChainRedeemMessageProcessor.scala

```diff
@@ -0,0 +1,150 @@
+  protected def processRedeemMessage(msg: AccountCrossChainRedeemMessage, view: BaseAccountStateView): Array[Byte] = {
+    validateRedeemMsg(msg, view)
+
+    addCrossChainMessageToView(view, msg)
+
+    addCrossChainRedeemMessageLogEvent(view, msg)
+
+    msg.encode()
+  }
+
+  private def addCrossChainRedeemMessageLogEvent(view: BaseAccountStateView, accCcRedeemMessage: AccountCrossChainRedeemMessage): Unit = {
+    val event = AddCrossChainRedeemMessage(
+      accCcRedeemMessage.accountCrossChainMessage,
+      accCcRedeemMessage.certificateDataHash,
+      accCcRedeemMessage.nextCertificateDataHash,
+      accCcRedeemMessage.scCommitmentTreeRoot,
+      accCcRedeemMessage.nextScCommitmentTreeRoot,
+      accCcRedeemMessage.proof
+    )
+    val evmLog = getEthereumConsensusDataLog(event)
+    view.addLog(evmLog)
+  }
+
+  private def addCrossChainMessageToView(view: BaseAccountStateView, accCcRedeemMessage: AccountCrossChainRedeemMessage): Unit = {
+    val messageHash = CryptoLibProvider.sc2scCircuitFunctions.getCrossChainMessageHash(
+      AbstractCrossChainMessageProcessor.buildCrosschainMessageFromAccount(accCcRedeemMessage.accountCrossChainMessage, networkParams)
+    )
+    setCrossChainMessageHash(messageHash, view)
+  }
+
+  /**
+   * Extract the AccountCrossChainRedeemMessage from the message
+   *
+   * @param msg message to apply to the state
+   * @return the AccountCrossChainRedeemMessage built from the arguments in the message
+   */
+  protected def getAccountCrossChainRedeemMessageFromMessage(msg: Message): AccountCrossChainRedeemMessage
+
+  private def validateRedeemMsg(ccRedeemMgs: AccountCrossChainRedeemMessage, view: BaseAccountStateView): Unit = {
+    try {
+      val accountCcMsg = ccRedeemMgs.accountCrossChainMessage
+      // validate arguments' semantics
+      validateSemantics(accountCcMsg.receiver, accountCcMsg.receiverSidechain, accountCcMsg.sender)
+
+      // Validate the receiving sidechain matches with this scId
+      validateScId(networkParams.sidechainId, accountCcMsg.receiverSidechain)
+
+      // Validate message has not been redeemed yet
+      val ccMsg = AbstractCrossChainMessageProcessor.buildCrosschainMessageFromAccount(accountCcMsg, networkParams)
+      validateDoubleMessageRedeem(ccMsg, view)
+
+      // Validate scCommitmentTreeRoot and nextScCommitmentTreeRoot exists
+      validateScCommitmentTreeRoot(ccRedeemMgs.scCommitmentTreeRoot, ccRedeemMgs.nextScCommitmentTreeRoot, view)
+
+      // Validate Proof
+      validateProof(ccRedeemMgs)
+    } catch {
+      case exception: IllegalArgumentException =>
+        throw new ExecutionRevertedException("Error while validating redeem message", exception)
+    }
+  }
+
+  private def validateSemantics(byteArraysToValidate: Array[Byte]*): Unit = {
+    val byteArrayMaxLength = 64
+    byteArraysToValidate.foreach(arrayToValidate => {
+      if (arrayToValidate.isEmpty || arrayToValidate.length > byteArrayMaxLength) {
```

Comment:
aren't we sure about the exact expected lengths?
like: sidechain id is strictly 32, receiver is exactly 20, etc.